### PR TITLE
tikv: fix region cache do not filter the down peers

### LINKED ISSUE #17336: region cache contain the down peers
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
use the follower read or tiflash read then make one peer down manaully.

### 2. What did you expect to see? (Required)
tidb don't access the down peer.

### 3. What did you see instead (Required)
there are lots of error logs in tidb-server, because it accesses the down peer.

### 4. Affected version (Required)
v4.0.0-RC.2

### 5. Root Cause Analysis
PD should tell tidb the down peers and tidb should filter it.

### PR DESCRIPTION
**Pull request by lzmhhh123**
### What problem does this PR solve?

Issue Number: close #17336

Problem Summary: filter the down peers to avoid accessing down peers in follower read and tiflash read(learner read).

### What is changed and how it works?

What's Changed: filter the available peers.

### Related changes

- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test
- Integration test

Side effects

- Performance regression
    - Consumes more CPU

### Release note 

- fix the problem tidb accesses the down peers in some cases.

### CODE REVIEW

File: store/tikv/region_cache.go [hunk 1]
```diff
@@ -831,6 +831,22 @@ func (c *RegionCache) getRegionByIDFromCache(regionID uint64) *Region {
...
+		for _, downPeer := range region.DownPeers {
+			if peer.Id == downPeer.Id && peer.StoreId == downPeer.StoreId {
+				available = false
+				break
+			}
+		}
+		if !available {
+			region.Meta.Peers = append(region.Meta.Peers[:i], region.Meta.Peers[i+1:]...)
+		}
+	}
```
Comment (bobotu) [inline store/tikv/region_cache.go:847]:
```suggestion
	if len(region.DownPeers) == 0 {
		return
	}
	new := region.Meta.Peers[:0]
	for _, p := range region.Meta.Peers {
		available := true
		for _, downPeer := range region.DownPeers {
			if peer.Id == downPeer.Id && peer.StoreId == downPeer.StoreId {
				available = false
				break
			}
		}
		if available {
			new = append(new, p)
		}
	}
	for i := len(new); i < len(region.Meta.Peers); i++ {
		region.Meta.Peers[i] = nil
	}
	region.Meta.Peers = new
```

What about the standard style of filtering slice?

Comment (lzmhhh123) [reply store/tikv/region_cache.go:847]: 👍

**Pull request merged**